### PR TITLE
Make @OpenShiftResource annotation work again with plain text

### LINKED ISSUE #1045: Cannot use plain text in @OpenShiftResource
##### Issue Overview

The JavaDoc says that it it possible to define the resource that should be deployed in OpenShift in the `@OpenShiftResource` annotation directly, but it's not true and creation of the resource fails.

##### Expected Behaviour

Resource is deployed.

##### Current Behaviour

```
	at org.arquillian.cube.openshift.impl.resources.OpenShiftResourceFactory.createResources(OpenShiftResourceFactory.java:76)
	at org.arquillian.cube.openshift.impl.CEEnvironmentProcessor.createEnvironment(CEEnvironmentProcessor.java:91)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:83)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:69)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:133)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:105)
	at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.beforeClass(EventTestRunnerAdaptor.java:89)
	at org.jboss.arquillian.junit.Arquillian$2.evaluate(Arquillian.java:163)
	at org.jboss.arquillian.junit.Arquillian.multiExecute(Arquillian.java:350)
	at org.jboss.arquillian.junit.Arquillian.access$200(Arquillian.java:54)
	at org.jboss.arquillian.junit.Arquillian$3.evaluate(Arquillian.java:177)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.jboss.arquillian.junit.Arquillian.run(Arquillian.java:115)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: java.lang.IllegalArgumentException: java.net.MalformedURLException: no protocol: {
  "apiVersion": "v1",
  "kind": "ServiceAccount",
  "metadata": {
    "name": "eap-service-account"
  }
}
	at org.arquillian.cube.kubernetes.impl.resolver.ResourceResolver.resolve(ResourceResolver.java:30)
	at org.arquillian.cube.openshift.impl.resources.OpenShiftResourceFactory.createOpenShiftResources(OpenShiftResourceFactory.java:116)
	at org.arquillian.cube.openshift.impl.resources.OpenShiftResourceFactory.createResources(OpenShiftResourceFactory.java:66)
	... 38 more
Caused by: java.net.MalformedURLException: no protocol: {
  "apiVersion": "v1",
  "kind": "ServiceAccount",
  "metadata": {
    "name": "eap-service-account"
  }
}
	at java.net.URL.<init>(URL.java:593)
	at java.net.URL.<init>(URL.java:490)
	at java.net.URL.<init>(URL.java:439)
	at org.arquillian.cube.kubernetes.impl.resolver.ResourceResolver.resolve(ResourceResolver.java:27)
	... 40 more
```

### PR DESCRIPTION
**Pull request by goldmann**
This proposal uses a temporary file workaround for the #1045 issue.

Fixes #1045

### CODE REVIEW

**Comment by lordofthejars** [comment]
Thinking about code, I am not sure if the best way is to do this or just create an interface `ResourceContent` which has a method called `getContent`.

Then we can have an implementation called `URLResourceContent` and `TextResourceContent`. The first one just do all `BufferReading` thingy to get an `String` from `URL` meanwhile the other one is just returning the String.

@dipak-pawar any thoughts about it?

**Comment by dipak-pawar** [comment]
@lordofthejars +1 for the suggestion which you proposed in above comment.

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resolver/ResourceResolver.java [hunk 1]
```diff
@@ -24,7 +27,20 @@ public static URL resolve(String location) {
 
                 return resource;
             } else {
-                return new URL(location);
+                // In case it is not a http, file nor a classpath protocol,
+                // we assume that this is plain text. We store it in a temporary
+                // file and return the URL to it.
+                File tmp = File.createTempFile("arquillian-cube", ".res");
```
Comment (lordofthejars) [inline kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resolver/ResourceResolver.java]: Can you move the creation of temporal file into a method like `createTemporalDefinition` which returns a `File` that is used later to create the URI? Just to have method more compact.

**Comment by goldmann** [comment]
@lordofthejars How about this?

**Comment by dipak-pawar** [comment]
@goldmann https://travis-ci.org/arquillian/arquillian-cube/jobs/388249666#L1213 Can you check this?

**Comment by lordofthejars** [comment]
@goldmann @dipak-pawar Are you receiving this compilation error in local too?

**Comment by dipak-pawar** [comment]
yes @lordofthejars, I am getting this locally too.

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resolver/ResourceResolver.java [hunk 2]
```diff
@@ -9,22 +11,22 @@
     static final String FILE_PREFIX = "file";
     static final String CLASSPATH_PREFIX = "classpath:";
 
-    public static URL resolve(String location) {
+    public static InputStream resolve(String content) {
```
Comment (dipak-pawar) [inline kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resolver/ResourceResolver.java]: This method is used in https://github.com/arquillian/arquillian-cube/blob/master/kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resources/KubernetesResourcesApplier.java#L102 which is expecting URL. You have to modify it to make it working as per modification

**Comment by goldmann** [comment]
Ooops, this one should be fixed now :)

**Comment by goldmann** [comment]
It looks like I'm reading some resources too early, hold on tight.

File: openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/utils/TemplateUtils.java [hunk 3]
```diff
@@ -72,7 +75,12 @@ public static String readTemplateUrl(Template template, CubeOpenShiftConfigurati
 
         if (templateUrl != null) {
             String url = resolver.resolve(templateUrl);
-            templateUrl = ResourceResolver.resolve(url).toString();
+
+            try {
+                templateUrl = new URL(url).toString();
+            } catch (MalformedURLException ex) {
+                throw new IllegalArgumentException(ex);
+            }
```
Comment (goldmann) [inline openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/utils/TemplateUtils.java]: The only reason we were "resolving" it earlier was to actually wrap it in URL and return string representation of it. Now we return `InputStream` in the `resolve` method reading it earlier. I decided to just validate the URL in the old form, as you can see.

**Comment by goldmann** [comment]
I'm giving up today, will start again tomorrow, forgot that it can be other things than url too :)

**Comment by goldmann** [comment]
Tests are failing today badly, because of timeout issues :(

**Comment by goldmann** [comment]
Phew, finally! I reverted to using the temporary file to return URL because returning `InputStream` was too problematic. I extracted preparing the file to separate method, as suggested.

**Pull request merged**